File: lab_1/i.py
def find_leader(n, votes):
    temp_list = list(votes)
    
    while len(set(temp_list)) > 1:
        i = 0
        while i < len(temp_list):
            if temp_list[i] == 'S':
                if 'K' in temp_list[i+1:]:
                    temp_list = temp_list[:i+1 + temp_list[i+1:].index('K')] + temp_list[i+2 + temp_list[i+1:].index('K'):]
                elif 'K' in temp_list:
                    temp_list = temp_list[:temp_list.index('K')] + temp_list[temp_list.index('K')+1:]
            elif temp_list[i] == 'K':
                if 'S' in temp_list[i+1:]:
                    temp_list = temp_list[:i+1 + temp_list[i+1:].index('S')] + temp_list[i+2 + temp_list[i+1:].index('S'):]
                elif 'S' in temp_list:
                    temp_list = temp_list[:temp_list.index('S')] + temp_list[temp_list.index('S')+1:]
            i += 1

    if temp_list[0] == 'S':
        return "SAKAYANAGI"
    else:
        return "KATSURAGI"

n = int(input().strip())
votes = input().strip()

print(find_leader(n, votes))

File: lab_1/i.cpp
#include <bits/stdc++.h>
using namespace std;

string solve(int n, string s) {
    deque<int> S, K;
    for (int i = 0; i < n; ++i) {
        (s[i] == 'S' ? S : K).push_back(i);
    }
    while (!S.empty() && !K.empty()) {
        if (S.front() < K.front()) {
            K.pop_front();
            S.push_back(S.front() + n);
            S.pop_front();
        } else {
            S.pop_front();
            K.push_back(K.front() + n);
            K.pop_front();
        }
    }
    return (S.empty() ? "KATSURAGI" : "SAKAYANAGI");
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int n;
    cin >> n;
    string votes;
    cin >> votes;
    
    cout << solve(n, votes) << endl;
    return 0;
}